### PR TITLE
Add material dictionary

### PR DESCRIPTION
**Pull request by coolsquid**
WIP - DO NOT MERGE

### CODE REVIEW

**Comment by iLexiconn** [comment]
Why can't you make the vanilla recipes accept the itemstacks?

**Comment by iLexiconn** [comment]
Format it properly

**Comment by iLexiconn** [comment]
Add the hashcode and equals to the original ItemStack and just get that one from the dictionary.

**Comment by shadowfacts** [comment]
@iLexiconn As I said before, `ItemStack`'s `#hashCode` and `#equals` will check item, meta, NBT, and stack size. The `ItemStackWrapper` ignores stack size, making it work properly with crafting recipes.

**Comment by iLexiconn** [comment]
Oh, just don't add stack size to the dictionary then.

**Comment by lclc98** [comment]
Item compare can be done using ItemStack.areItemsEqual

File: src/main/java/xyz/openmodloader/dictionary/ItemStackWrapper.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * Instantiates a new item stack wrapper.
+     *
+     * @param stack the stack - can be null
+     */
+    public ItemStackWrapper(ItemStack stack) {
+        this.stack = stack;
+    }
+
```
Comment (Victorious3) [inline src/main/java/xyz/openmodloader/dictionary/ItemStackWrapper.java]: Don't let eclipse auto generate these, use `Objects.hash(...)`

File: src/main/java/xyz/openmodloader/dictionary/ItemStackWrapper.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            return false;
+        if (stack.getItemDamage() != other.getItemDamage())
+            return false;
+        if (stack.getTagCompound() == null) {
+            if (other.getTagCompound() != null)
+                return false;
+        } else if (!stack.getTagCompound().equals(other.getTagCompound()))
+            return false;
+        return true;
+    }
```
Comment (Victorious3) [inline src/main/java/xyz/openmodloader/dictionary/ItemStackWrapper.java]: Same as above, those should always be written by hand

File: src/main/java/xyz/openmodloader/dictionary/ItemStackWrapper.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+import net.minecraft.item.ItemStack;
+
+/**
+ * A wrapper around ItemStack that contains
+ * versions of {@link #hashCode()} and {@link #equals(Object)}
+ * that take the item, the item damage, and the NBT into
+ * account, but not the stack size.
+ */
+public class ItemStackWrapper {
```
Comment (Victorious3) [inline src/main/java/xyz/openmodloader/dictionary/ItemStackWrapper.java]: Is this whole class needed at all? Remember, we can override hashCode and equals for `ItemStack`, this isn't Forge

Comment (iLexiconn) [inline src/main/java/xyz/openmodloader/dictionary/ItemStackWrapper.java]: What I said too ^

**Pull request merged**